hub/bolt_transport_test.go

```diff
@@ -0,0 +1,247 @@
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	bolt "go.etcd.io/bbolt"
+)
+
+func TestBoltTransportHistory(t *testing.T) {
+	url, _ := url.Parse("bolt://test.db")
+	transport, _ := NewBoltTransport(&Options{TransportURL: url})
+	defer transport.Close()
+	defer os.Remove("test.db")
+
+	for i := 1; i <= 10; i++ {
+		transport.Write(&Update{Event: Event{ID: strconv.Itoa(i)}})
+	}
+
+	pipe, err := transport.CreatePipe("8")
+	assert.Nil(t, err)
+	require.NotNil(t, pipe)
+
+	var count uint64
+	go func() {
+		for {
+			_, err := pipe.Read(context.Background())
+			if err == ErrClosedPipe {
+				return
+			}
+			atomic.AddUint64(&count, 1)
+		}
+	}()
+
+	// let time to the reading loop to process as many message as it can. Then we close the pipe
+	time.Sleep(10 * time.Millisecond)
+	pipe.Close()
+
+	// the reading loop should have read the #9 and #10 messages
+	assert.Equal(t, uint64(2), atomic.LoadUint64(&count))
+}
+
+func TestBoltTransportHistoryAndLive(t *testing.T) {
+	url, _ := url.Parse("bolt://test.db")
+	transport, _ := NewBoltTransport(&Options{TransportURL: url})
+	defer transport.Close()
+	defer os.Remove("test.db")
+
+	for i := 1; i <= 10; i++ {
+		transport.Write(&Update{Event: Event{ID: strconv.Itoa(i)}})
+	}
+
+	pipe, err := transport.CreatePipe("8")
+	assert.Nil(t, err)
+	require.NotNil(t, pipe)
+
+	var count uint64
+	go func() {
+		for {
+			_, err := pipe.Read(context.Background())
+			if err == ErrClosedPipe {
+				return
+			}
+			atomic.AddUint64(&count, 1)
+		}
+	}()
+
+	transport.Write(&Update{Event: Event{ID: "11"}})
+
+	// let time to the reading loop to process as many message as it can. Then we close the pipe
+	time.Sleep(10 * time.Millisecond)
+	pipe.Close()
+
+	// the reading loop should have read the #9, #10 messages then the #11
+	assert.Equal(t, uint64(3), atomic.LoadUint64(&count))
+}
+
+func TestBoltTransportPurgeHistory(t *testing.T) {
+	url, _ := url.Parse("bolt://test.db?size=5&cleanup_frequency=1")
+	transport, _ := NewBoltTransport(&Options{TransportURL: url})
+	defer transport.Close()
+	defer os.Remove("test.db")
+
+	for i := 0; i < 12; i++ {
+		transport.Write(&Update{Event: Event{ID: strconv.Itoa(i)}})
+	}
+
+	transport.db.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte("updates"))
+
+		assert.Equal(t, 5, b.Stats().KeyN)
+
+		return nil
+	})
+}
+
+func TestNewBoltTransport(t *testing.T) {
+	url, _ := url.Parse("bolt://test.db?bucket_name=demo")
+	transport, err := NewBoltTransport(&Options{TransportURL: url})
+	assert.Nil(t, err)
+	require.NotNil(t, transport)
+	transport.Close()
+
+	url, _ = url.Parse("bolt://")
+	_, err = NewBoltTransport(&Options{TransportURL: url})
+	assert.EqualError(t, err, `invalid bolt "bolt:" dsn: missing path`)
+
+	url, _ = url.Parse("bolt:///root/test.db")
+	_, err = NewBoltTransport(&Options{TransportURL: url})
+	assert.EqualError(t, err, `invalid bolt "bolt:///root/test.db" dsn: open /root/test.db: permission denied`)
```

Comment:
This test fail on Mac OS:

        	Error:      	Error message not equal:
        	            	expected: "invalid bolt \"bolt:///root/test.db\" dsn: open /root/test.db: permission denied"
        	            	actual  : "invalid bolt \"bolt:///root/test.db\" dsn: open /root/test.db: no such file or directory"

Reply:
Using `/` instead of `/root` fixes the issue.